### ISSUE TITLE
Incorrect unused NamedFieldPuns language pragma warning

### ISSUE THREAD

**Issue by dten** [ndmitchell/hlint#767]
If module A imports a datatype and its fields from module B, and module A uses NamedFieldPuns and has a binding with a name the same as a field from A then NamedFieldPuns is warned as unused, but it actually is necessary

**Comment by ndmitchell** [comment]
Do you have a small example? Is the NamedFieldPun shadowing the binding in A? I'm struggling to visualise this precisely.

**Comment by dten** [comment]
```
module A where
data MyType { myField :: Int }
```
```
module B where
import A
myFn = do
   let myField = 1
   return myField
```

**Comment by dten** [comment]
extended the example a little to make sure it compiles even with `-Wall`  (which we have on)

**Comment by ndmitchell** [comment]
I was able to remove `NamedFieldPuns` from B and it still worked. What error do you get?

**Comment by dten** [comment]
Try it with `-Wall` you only need puns for the would be shadowing error

**Comment by ndmitchell** [comment]
OK, now I can reproduce. What I don't understand is why having NamedFieldPuns causes this not to be an error? Why isn't this a GHC bug?

**Comment by dten** [comment]
Hmmm well I didn't write the original code that added `NamedFieldPuns` I just updated our hlint. Assumed it was an intended side effect of the extension that it doesn't count as shadowing.

**Comment by dten** [comment]
https://gitlab.haskell.org/ghc/ghc/issues/14630#note_148282

There's a bit of discussion here which implies this and record wildcards is to aggressive in stunning things aren't shadowed.

I ended up removing the pragma and fixing the shadowing. Tempted to say that the correct solution and to close this

**Comment by ndmitchell** [comment]
Yep, to my mind this is a GHC bug - the presence of NamedFieldPuns makes the warning checker just give up, for no good reason in this case. Closing.
